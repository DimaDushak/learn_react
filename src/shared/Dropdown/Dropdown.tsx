/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
import React from 'react';
import { useIsMounted } from '../../utils/hooks';
import styles from './dropdown.css';

interface Props {
    button: React.ReactNode;
    children: React.ReactNode;
    listClassName?: string;
    isOpen?: boolean;
    onOpen?: () => void;
    onClose?: () => void;
}

export default function Dropdown({
    button,
    children,
    listClassName = styles.listPosition,
    isOpen,
    onOpen,
    onClose,
}: Props): JSX.Element {
    const [isDropdownOpen, setIsDropDownOpen] = React.useState(isOpen);

    const isMounted = useIsMounted();

    React.useEffect(() => setIsDropDownOpen(isOpen), [isOpen]);

    React.useEffect(() => {
        if (isMounted) {
            // eslint-disable-next-line @typescript-eslint/no-unused-expressions
            isDropdownOpen ? onOpen?.() : onClose?.();
        }
    }, [isDropdownOpen]);

    const handleOpen = () => {
        if (isOpen === undefined) {
            setIsDropDownOpen((prevState) => !prevState);
        }
    };

    return (
        <div className={styles.container}>
            <div onClick={handleOpen}>{button}</div>
            {isDropdownOpen && (
                <div className={styles.listContainer}>
                    <div
                        className={`${listClassName} ${styles.list}`}
                        onClick={() => setIsDropDownOpen(false)}
                    >
                        {children}
                    </div>
                </div>
            )}
        </div>
    );
}
