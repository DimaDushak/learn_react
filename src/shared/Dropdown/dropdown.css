.container {
    flex: 1;
    width: 100%;
}

.listContainer {
    position: relative;
}

.list {
    height: 100%;
    position: absolute;
    width: 100%;
}

.listPosition {
    left: 0;
    top: 0;
}
