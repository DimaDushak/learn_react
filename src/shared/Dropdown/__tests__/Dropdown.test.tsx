import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import Dropdown from '../Dropdown';

describe('Dropdown', () => {
    test('should render (snapshot)', () => {
        const { container } = render(
            <Dropdown button={<button type="button">Нажми</button>}>
                <ul>Список</ul>
            </Dropdown>,
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    test('render after button click (snapshot)', () => {
        const { container } = render(
            <Dropdown button={<button type="button">Нажми</button>}>
                <ul>Список</ul>
            </Dropdown>,
        );

        userEvent.click(screen.getByRole('button'));

        expect(container.firstChild).toMatchSnapshot();
    });

    test('should work', () => {
        render(
            <Dropdown button={<button type="button">Нажми</button>}>
                <ul>Список</ul>
            </Dropdown>,
        );

        expect(screen.queryByRole('list')).toBeNull();

        const button = screen.getByRole('button');

        userEvent.click(button);

        expect(screen.getByRole('list')).toBeInTheDocument();

        userEvent.click(button);

        expect(screen.queryByRole('list')).toBeNull();

        userEvent.click(button);
        userEvent.click(screen.getByRole('list'));

        expect(screen.queryByRole('list')).toBeNull();
    });

    test('should work controlled (check prop isOpen)', () => {
        render(
            <Dropdown
                button={<button type="button">Нажми</button>}
                isOpen={false}
            >
                <ul>Список</ul>
            </Dropdown>,
        );

        userEvent.click(screen.getByRole('button'));

        expect(screen.queryByRole('list')).toBeNull();
    });

    test('shoud work controlled (check props onOpen and onClose)', () => {
        const onOpen = jest.fn();
        const onClose = jest.fn();

        render(
            <Dropdown
                button={<button type="button">Нажми</button>}
                onOpen={onOpen}
                onClose={onClose}
            >
                <ul>Список</ul>
            </Dropdown>,
        );

        const button = screen.getByRole('button');

        expect(onClose).toBeCalledTimes(0);

        userEvent.click(button);

        expect(onOpen).toBeCalled();

        userEvent.click(button);

        expect(onClose).toBeCalledTimes(1);
    });

    test('check prop listClassName', () => {
        render(
            <Dropdown
                button={<button type="button">Нажми</button>}
                listClassName="changedlistPosition"
                isOpen
            >
                <ul>Список</ul>
            </Dropdown>,
        );

        expect(screen.getByRole('list').parentElement).toHaveClass(
            'changedlistPosition',
        );
    });
});
