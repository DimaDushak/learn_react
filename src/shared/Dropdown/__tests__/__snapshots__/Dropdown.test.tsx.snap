// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown render after button click (snapshot) 1`] = `
<div
  class="container"
>
  <div>
    <button
      type="button"
    >
      Нажми
    </button>
  </div>
  <div
    class="listContainer"
  >
    <div
      class="listPosition list"
    >
      <ul>
        Список
      </ul>
    </div>
  </div>
</div>
`;

exports[`Dropdown should render (snapshot) 1`] = `
<div
  class="container"
>
  <div>
    <button
      type="button"
    >
      Нажми
    </button>
  </div>
</div>
`;
